### ISSUE TITLE
Can't load metadata from **both** filename and file contents

### ISSUE THREAD

**Issue by sharat87** [getnikola/nikola#3344]
# Environment

**Python Version:** 3.8
**Nikola Version:** 8.0.3
**Operating System:** Windows 7

# Description:

I created a new website with Nikola 8.0.3, trying to move my blog that is currently built with a custom script. My blog posts are written in Markdown files with names like `'2019-12-31_post-slug-goes-here.md'`. The file would contain additional metadata like title, description, whether it should have comments etc. as a YAML block at the start, delimited by `---`.

From the documentation and trying to dig the source code, I can see the Nikola supports both of these styles of loading metadata, *individually*, but not both. So, with a file like I mentioned above, Nikola would load the metadata from the YAML block and complain that the post doesn't have a date. If I remove the YAML block inside, it loads the metadata from the filename and happily continues to build. But of course, it would be lacking post title, description etc.

My question is, is it possible to have Nikola load metadata from the filename first and then from the YAML block, overwriting anything that repeats. The order of loading from the two sources could be anything (I'm indifferent). I just want it to load metadata from *both* the sources.

I see from [this line in post.py](https://github.com/getnikola/nikola/blob/bcc8dbcc125185566338b1228674faad1459e5b2/nikola/post.py#L1039) that filename metadata is treated as fallback and is *only* loaded if there's no metadata in the content. Wouldn't it be better to load filename metadata if the `FILE_METADATA_REGEXP` option is set, irrespective of whether the file content has metadata or not? Please disregard if I'm missing the reason behind this choice.

Thank you very much for your work on Nikola. Also, I've searched issues but couldn't find this being discussed before. Sorry if this is a duplicate.

**Comment by sharat87** [comment]
I just tested by changing the line `if not meta:` on line 1040 to `if config.get('FILE_METADATA_REGEXP'):` in my local copy of Nikola and it's working as I wanted it to. I first loads metadata from the file's content and then updates the metadata dictionary with what's loaded from the filename.

I can make a PR if it sounds acceptable. Thanks!

**Comment by ralsina** [comment]
Sounds good to me.

**Comment by Kwpolska** [comment]
Filename-based metadata is currently considered a fallback, because in general, if you can put Nikola-specific (or Nikola-accepted) metadata inside the file, you can probably put *all* metadata in that file, and I’d consider that the more convenient option where everything is in one place.

> I first loads metadata from the file's content and then updates the metadata dictionary with what's loaded from the filename.

I’m indifferent about it in general, but that should work the other way around, with file content being more important than the filename.

**Comment by sharat87** [comment]
> I’d consider that the more convenient option where everything is in one place.

Yeah, I can agree with that sentiment. My reason to put only the date and slug in the filename is so that the files are all sorted by date and it doesn't take any effort to locate the latest post. The file names usually have the same character limitations as slugs, so I was loading the slug from the filename. So, even if I wanted to keep all metadata at one place, I was anyway saving the date and slug to the filename. That was how I ended up with what I have today :).

> [...] with file content being more important than the filename.

Sure. In that case, the PR would be more involved. The change I suggested in my previous comment wouldn't do it in this order. It's late here so I'll hopefully spend some time tomorrow.
